exporter/metric/prometheus/prometheus.go

```diff
@@ -0,0 +1,188 @@
+import (
+	"context"
+	"net/http"
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+
+	"go.opentelemetry.io/otel/api/core"
+	export "go.opentelemetry.io/otel/sdk/export/metric"
+	"go.opentelemetry.io/otel/sdk/export/metric/aggregator"
+	"go.opentelemetry.io/otel/sdk/metric"
+)
+
+type metricKey struct {
+	desc    *export.Descriptor
+	encoded string
+}
+
+// Exporter is an implementation of metric.Exporter that sends metrics to
+// Prometheus.
+type Exporter struct {
+	sync.RWMutex
+
+	labelEncoder export.LabelEncoder
+	handler      http.Handler
+	registerer   prometheus.Registerer
+	gatherer     prometheus.Gatherer
+	counters     counters
+	gauges       gauges
+	histograms   histograms
+}
+
+var _ export.Exporter = &Exporter{}
+var _ http.Handler = &Exporter{}
+var _ export.LabelEncoder = &Exporter{}
+
+// Options is a set of options for the tally reporter.
+type Options struct {
+	// Registry is the prometheus registry that will be used as the default Registerer and
+	// Gatherer if these are not specified.
+	//
+	// If not set a new empty Registry is created.
+	Registry *prometheus.Registry
+
+	// Registerer is the prometheus registerer to register
+	// metrics with.
+	//
+	// If not specified the Registry will be used as default.
+	Registerer prometheus.Registerer
+
+	// Gatherer is the prometheus gatherer to gather
+	// metrics with.
+	//
+	// If not specified the Registry will be used as default.
+	Gatherer prometheus.Gatherer
+
+	// DefaultHistogramBuckets is the default histogram buckets
+	// to use. Use nil to specify the system-default histogram buckets.
+	DefaultHistogramBuckets []float64
+
+	// LabelEncoder is the label encoder that will be used to group
+	// and export metrics to Prometheus.
+	LabelEncoder export.LabelEncoder
+}
+
+// NewExporter returns a new prometheus exporter for prometheus metrics.
+func NewExporter(opts Options) (*Exporter, error) {
+	if opts.Registry == nil {
+		opts.Registry = prometheus.NewRegistry()
+	}
+
+	if opts.Registerer == nil {
+		opts.Registerer = opts.Registry
+	}
+
+	if opts.Gatherer == nil {
+		opts.Gatherer = opts.Registry
+	}
+
+	if opts.LabelEncoder == nil {
+		opts.LabelEncoder = metric.NewDefaultLabelEncoder()
+	}
+
+	return &Exporter{
+		labelEncoder: opts.LabelEncoder,
+
+		registerer: opts.Registerer,
+		gatherer:   opts.Gatherer,
+		handler:    promhttp.HandlerFor(opts.Gatherer, promhttp.HandlerOpts{}),
+
+		counters:   newCounters(opts.Registerer),
+		gauges:     newGauges(opts.Registerer),
+		histograms: newHistograms(opts.Registerer, opts.DefaultHistogramBuckets),
+	}, nil
+}
+
+// Export exports the provide metric record to prometheus.
+func (e *Exporter) Export(_ context.Context, checkpointSet export.CheckpointSet) error {
+	var forEachError error
+	checkpointSet.ForEach(func(record export.Record) {
+		agg := record.Aggregator()
+		labels := record.Labels()
+
+		var encoded string
+		if labels.Encoder() == e {
```

Comment:
It's not clear to me that this matters. Any label encoder should uphold the contract, which is that encodings are unique. I'm not sure it needs to be part of the options either. If we're not using the encoded form directly for export (if it's only a map key), any encoder should do and we should just take the encoding we get.

Reply:
Yeah, I've mixed things up.

Reply:
Now that I'm thinking about it. We talked about parsing the `encoded` labels instead of calling `Emit` on the `labelValues` method, to make that we would need to use a specific encoding. So I guess we would need this and remove option for changing the encoding or dropping this and the `TODO` about parsing labels.


For reference:
```go
func labelValues(labels export.Labels) []string {
	// TODO(paivagustavo): parse the labels.Encoded() instead of calling `Emit()` directly
	//  this would avoid unnecessary allocations.
	values := make([]string, 0, labels.Len())
	for _, label := range labels.Ordered() {
		values = append(values, label.Value.Emit())
	}
	return values
}
```